File: kernel_pacer_jvm/src/main/java/com/xa/mass/kernel/pacer/assignment/DispatchConvergenceApplication.java
package com.xa.mass.kernel.pacer;

import java.time.Duration;
import java.util.ArrayList;
import java.util.EnumMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.RejectedExecutionException;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.TimeUnit;

final class DispatchConvergenceApplication {

    private static final System.Logger LOGGER = System.getLogger(
            DispatchConvergenceApplication.class.getName()
    );

    private final Object lifecycleLock = new Object();
    private final TaskSchedulingBatchSource source;
    private final TaskRunningActivationPolicy activation;
    private final TaskWorkerAllocationPolicy allocation;
    private final TaskDispatchPolicy dispatch;
    private final WorkerServiceabilityDispatchPolicy serviceability;
    private Thread coordinator;
    private ExecutorService batchExecutor;
    private CountDownLatch stopSignal;
    private State state = State.STOPPED;

    DispatchConvergenceApplication(
            TaskSchedulingBatchSource source,
            TaskRunningActivationPolicy activation,
            TaskWorkerAllocationPolicy allocation,
            TaskDispatchPolicy dispatch,
            WorkerServiceabilityDispatchPolicy serviceability
    ) {
        this.source = Objects.requireNonNull(source, "source");
        this.activation = Objects.requireNonNull(activation, "activation");
        this.allocation = Objects.requireNonNull(allocation, "allocation");
        this.dispatch = Objects.requireNonNull(dispatch, "dispatch");
        this.serviceability = serviceability;
    }

    void start(
            AssignmentDispatchConfig assignmentConfig,
            WorkerServiceabilityAssemblyConfig serviceabilityConfig
    ) {
        Objects.requireNonNull(assignmentConfig, "assignmentConfig");
        Objects.requireNonNull(serviceabilityConfig, "serviceabilityConfig");
        synchronized (lifecycleLock) {
            if (coordinator != null || state != State.STOPPED) {
                throw new IllegalStateException(
                        "Dispatch Convergence application is already started"
                );
            }
            CountDownLatch signal = new CountDownLatch(1);
            BlockingQueue<LaneCompletion> completions =
                    new LinkedBlockingQueue<>();
            ThreadFactory batchThreads = Thread.ofVirtual()
                    .name("dispatch-convergence-batch-", 0)
                    .factory();
            ExecutorService executor = Executors.newThreadPerTaskExecutor(
                    batchThreads
            );
            List<LaneDefinition> lanes = lanes(
                    assignmentConfig,
                    serviceabilityConfig
            );
            Thread started = new Thread(
                    () -> runLoop(signal, completions, executor, lanes),
                    "dispatch-convergence-coordinator"
            );
            started.setDaemon(false);
            stopSignal = signal;
            batchExecutor = executor;
            coordinator = started;
            state = State.RUNNING;
            started.start();
        }
    }

    void stop(long timeoutMillis) {
        if (timeoutMillis < 1) {
            throw new IllegalArgumentException(
                    "timeoutMillis must be positive"
            );
        }
        Thread current;
        ExecutorService executor;
        CountDownLatch signal;
        synchronized (lifecycleLock) {
            current = coordinator;
            executor = batchExecutor;
            signal = stopSignal;
            if (current == null || executor == null || signal == null) {
                return;
            }
            state = State.STOPPING;
            signal.countDown();
            current.interrupt();
        }
        long deadline = System.nanoTime()
                + Duration.ofMillis(timeoutMillis).toNanos();
        join(current, remainingMillis(deadline));
        if (current.isAlive()) {
            executor.shutdownNow();
            failStoppedState();
            throw new IllegalStateException(
                    "Dispatch Convergence coordinator did not stop within "
                            + "its budget"
            );
        }
        if (!awaitTermination(executor, remainingMillis(deadline))) {
            executor.shutdownNow();
            failStoppedState();
            throw new IllegalStateException(
                    "Dispatch Convergence batches did not stop within "
                            + "their budget"
            );
        }
        synchronized (lifecycleLock) {
            if (coordinator == current) {
                coordinator = null;
                batchExecutor = null;
                stopSignal = null;
                state = State.STOPPED;
            }
        }
    }

    boolean isRunning() {
        synchronized (lifecycleLock) {
            refreshDeadCoordinator();
            return state == State.RUNNING
                    && coordinator != null
                    && coordinator.isAlive()
                    && batchExecutor != null
                    && !batchExecutor.isShutdown();
        }
    }

    String state() {
        synchronized (lifecycleLock) {
            refreshDeadCoordinator();
            return state.name();
        }
    }

    private List<LaneDefinition> lanes(
            AssignmentDispatchConfig assignmentConfig,
            WorkerServiceabilityAssemblyConfig serviceabilityConfig
    ) {
        List<LaneDefinition> definitions = new ArrayList<>();
        definitions.add(LaneDefinition.fromMillis(
                DispatchLaneId.RUNNING_ACTIVATION,
                assignmentConfig.runningActivationIntervalMillis(),
                batch -> activation.activateRunningVisibleTasks(
                        batch,
                        assignmentConfig.runningActivation()
                )
        ));
        definitions.add(LaneDefinition.fromMillis(
                DispatchLaneId.WORKER_ALLOCATION,
                assignmentConfig.workerAllocationIntervalMillis(),
                batch -> allocation.allocateCandidateWorkers(
                        batch,
                        assignmentConfig.workerAllocation()
                )
        ));
        definitions.add(LaneDefinition.fromMillis(
                DispatchLaneId.TASK_DISPATCH,
                assignmentConfig.taskDispatchIntervalMillis(),
                batch -> dispatch.dispatchTasks(
                        batch,
                        assignmentConfig.taskDispatch()
                )
        ));
        if (serviceabilityConfig.enabled()) {
            if (serviceability == null) {
                throw new IllegalStateException(
                        "enabled serviceability requires its dispatch policy"
                );
            }
            definitions.add(LaneDefinition.fromMillis(
                    DispatchLaneId.WORKER_SERVICEABILITY,
                    serviceabilityConfig.dispatch().intervalMillis(),
                    batch -> serviceability.dispatchProbes(
                            batch,
                            serviceabilityConfig.dispatch().dispatch(),
                            serviceabilityConfig.hotEligibilityFloorMillis()
                    )
            ));
        }
        return List.copyOf(definitions);
    }

    private void runLoop(
            CountDownLatch signal,
            BlockingQueue<LaneCompletion> completions,
            ExecutorService executor,
            List<LaneDefinition> lanes
    ) {
        Map<DispatchLaneId, LaneRuntime> runtimes = new EnumMap<>(
                DispatchLaneId.class
        );
        lanes.forEach(lane -> runtimes.put(
                lane.id(),
                new LaneRuntime(lane)
        ));
        boolean stopping = false;
        try {
            while (signal.getCount() > 0) {
                drainCompletions(runtimes, completions);
                dispatchRunningBatch(
                        runtimes,
                        completions,
                        executor,
                        signal
                );
                dispatchAdmissionBatch(
                        runtimes,
                        completions,
                        executor,
                        signal
                );
                if (signal.getCount() == 0) {
                    stopping = true;
                    break;
                }
                LaneCompletion completion = waitForWork(
                        runtimes,
                        completions
                );
                if (completion != null) {
                    applyCompletion(runtimes, completion);
                }
            }
            stopping = true;
        } catch (InterruptedException interrupted) {
            if (signal.getCount() == 0) {
                stopping = true;
            } else {
                Thread.currentThread().interrupt();
                throw new IllegalStateException(
                        "Dispatch Convergence coordinator was interrupted",
                        interrupted
                );
            }
        } finally {
            if (stopping) {
                executor.shutdown();
            } else {
                executor.shutdownNow();
            }
            synchronized (lifecycleLock) {
                if (coordinator == Thread.currentThread()
                        && state != State.STOPPING) {
                    state = State.FAILED;
                }
            }
        }
    }

    private void dispatchRunningBatch(
            Map<DispatchLaneId, LaneRuntime> runtimes,
            BlockingQueue<LaneCompletion> completions,
            ExecutorService executor,
            CountDownLatch signal
    ) {
        long now = System.nanoTime();
        List<LaneRuntime> eligible = runtimes.values().stream()
                .filter(runtime -> runtime.lane.id()
                        != DispatchLaneId.RUNNING_ACTIVATION)
                .filter(runtime -> runtime.idleAndEligible(now))
                .toList();
        if (eligible.isEmpty() || signal.getCount() == 0) {
            return;
        }
        List<DueTaskObservation> batch;
        try {
            batch = source.acquireRunningTasks(
                    AssignmentDispatchConfig.TASK_BATCH_LIMIT
            );
        } catch (RuntimeException failure) {
            eligible.forEach(this::deferLane);
            logFailure("runningSource", null, 0, failure);
            return;
        }
        if (batch.isEmpty()) {
            eligible.forEach(this::deferLane);
            return;
        }
        eligible.forEach(runtime -> submit(
                runtime,
                batch,
                completions,
                executor
        ));
    }

    private void dispatchAdmissionBatch(
            Map<DispatchLaneId, LaneRuntime> runtimes,
            BlockingQueue<LaneCompletion> completions,
            ExecutorService executor,
            CountDownLatch signal
    ) {
        LaneRuntime runtime = runtimes.get(
                DispatchLaneId.RUNNING_ACTIVATION
        );
        if (runtime == null
                || !runtime.idleAndEligible(System.nanoTime())
                || signal.getCount() == 0) {
            return;
        }
        List<DueTaskObservation> batch;
        try {
            batch = source.acquireAdmissionTasks(
                    AssignmentDispatchConfig.TASK_BATCH_LIMIT
            );
        } catch (RuntimeException failure) {
            deferLane(runtime);
            logFailure("admissionSource", runtime.lane.id(), 0, failure);
            return;
        }
        if (batch.isEmpty()) {
            deferLane(runtime);
            return;
        }
        submit(runtime, batch, completions, executor);
    }

    private void submit(
            LaneRuntime runtime,
            List<DueTaskObservation> batch,
            BlockingQueue<LaneCompletion> completions,
            ExecutorService executor
    ) {
        runtime.inflight = true;
        try {
            executor.submit(() -> executeBatch(
                    runtime.lane,
                    batch,
                    completions
            ));
        } catch (RejectedExecutionException failure) {
            runtime.inflight = false;
            throw new IllegalStateException(
                    "Dispatch Convergence executor rejected lane="
                            + runtime.lane.id(),
                    failure
            );
        }
    }

    private static void executeBatch(
            LaneDefinition lane,
            List<DueTaskObservation> batch,
            BlockingQueue<LaneCompletion> completions
    ) {
        Throwable failure = null;
        try {
            lane.policy().handle(batch);
        } catch (RuntimeException runtimeFailure) {
            failure = runtimeFailure;
        } catch (Error fatalFailure) {
            failure = fatalFailure;
            throw fatalFailure;
        } finally {
            completions.offer(new LaneCompletion(
                    lane.id(),
                    batch.size(),
                    failure
            ));
        }
    }

    private void drainCompletions(
            Map<DispatchLaneId, LaneRuntime> runtimes,
            BlockingQueue<LaneCompletion> completions
    ) {
        LaneCompletion completion;
        while ((completion = completions.poll()) != null) {
            applyCompletion(runtimes, completion);
        }
    }

    private void applyCompletion(
            Map<DispatchLaneId, LaneRuntime> runtimes,
            LaneCompletion completion
    ) {
        LaneRuntime runtime = runtimes.get(completion.id());
        if (runtime == null || !runtime.inflight) {
            throw new IllegalStateException(
                    "Unexpected Dispatch lane completion: "
                            + completion.id()
            );
        }
        runtime.inflight = false;
        deferLane(runtime);
        if (completion.failure() == null) {
            return;
        }
        if (completion.failure() instanceof Error fatal) {
            throw fatal;
        }
        logFailure(
                "policy",
                runtime.lane.id(),
                completion.batchSize(),
                (RuntimeException) completion.failure()
        );
    }

    private LaneCompletion waitForWork(
            Map<DispatchLaneId, LaneRuntime> runtimes,
            BlockingQueue<LaneCompletion> completions
    ) throws InterruptedException {
        long now = System.nanoTime();
        long waitNanos = Long.MAX_VALUE;
        boolean inflight = false;
        for (LaneRuntime runtime : runtimes.values()) {
            if (runtime.inflight) {
                inflight = true;
                continue;
            }
            waitNanos = Math.min(
                    waitNanos,
                    Math.max(0, runtime.nextEligibleNanos - now)
            );
        }
        if (waitNanos == 0) {
            return null;
        }
        if (waitNanos == Long.MAX_VALUE) {
            return inflight ? completions.take() : null;
        }
        return completions.poll(waitNanos, TimeUnit.NANOSECONDS);
    }

    private void deferLane(LaneRuntime runtime) {
        runtime.nextEligibleNanos = Math.addExact(
                System.nanoTime(),
                runtime.lane.intervalNanos()
        );
    }

    private static void logFailure(
            String operation,
            DispatchLaneId lane,
            int batchSize,
            RuntimeException failure
    ) {
        LOGGER.log(
                System.Logger.Level.ERROR,
                "operation=dispatchConvergence.{0} lane={1} batchSize={2} "
                        + "failureType={3}",
                operation,
                lane,
                batchSize,
                failure.getClass().getName()
        );
    }

    private static void join(Thread thread, long timeoutMillis) {
        try {
            thread.join(timeoutMillis);
        } catch (InterruptedException interrupted) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException(
                    "Dispatch Convergence shutdown was interrupted",
                    interrupted
            );
        }
    }

    private static boolean awaitTermination(
            ExecutorService executor,
            long timeoutMillis
    ) {
        try {
            return executor.awaitTermination(
                    timeoutMillis,
                    TimeUnit.MILLISECONDS
            );
        } catch (InterruptedException interrupted) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException(
                    "Dispatch Convergence batch shutdown was interrupted",
                    interrupted
            );
        }
    }

    private static long remainingMillis(long deadlineNanos) {
        return Math.max(
                1,
                Duration.ofNanos(Math.max(
                        1,
                        deadlineNanos - System.nanoTime()
                )).toMillis()
        );
    }

    private void failStoppedState() {
        synchronized (lifecycleLock) {
            state = State.FAILED;
        }
    }

    private void refreshDeadCoordinator() {
        if (state == State.RUNNING
                && coordinator != null
                && !coordinator.isAlive()) {
            state = State.FAILED;
        }
    }

    private enum State {
        STOPPED,
        RUNNING,
        STOPPING,
        FAILED
    }

    private enum DispatchLaneId {
        RUNNING_ACTIVATION,
        WORKER_ALLOCATION,
        TASK_DISPATCH,
        WORKER_SERVICEABILITY
    }

    @FunctionalInterface
    private interface DispatchBatchPolicy {
        void handle(List<DueTaskObservation> batch);
    }

    private record LaneDefinition(
            DispatchLaneId id,
            long intervalNanos,
            DispatchBatchPolicy policy
    ) {
        private LaneDefinition {
            Objects.requireNonNull(id, "id");
            Objects.requireNonNull(policy, "policy");
            if (intervalNanos < 1) {
                throw new IllegalArgumentException(
                        "Dispatch lane interval must be positive"
                );
            }
        }

        private static LaneDefinition fromMillis(
                DispatchLaneId id,
                long intervalMillis,
                DispatchBatchPolicy policy
        ) {
            return new LaneDefinition(
                    Objects.requireNonNull(id, "id"),
                    TimeUnit.MILLISECONDS.toNanos(requirePositive(
                            intervalMillis
                    )),
                    Objects.requireNonNull(policy, "policy")
            );
        }

        private static long requirePositive(long value) {
            if (value < 1) {
                throw new IllegalArgumentException(
                        "Dispatch lane interval must be positive"
                );
            }
            return value;
        }
    }

    private static final class LaneRuntime {
        private final LaneDefinition lane;
        private boolean inflight;
        private long nextEligibleNanos;

        private LaneRuntime(LaneDefinition lane) {
            this.lane = lane;
        }

        private boolean idleAndEligible(long nowNanos) {
            return !inflight && nowNanos >= nextEligibleNanos;
        }
    }

    private record LaneCompletion(
            DispatchLaneId id,
            int batchSize,
            Throwable failure
    ) {
    }
}

File: kernel_design/executable_spec/assembly/dispatch_convergence_application.py
from __future__ import annotations

import logging
from dataclasses import dataclass
from enum import Enum
from queue import Empty, Queue
from threading import Event, Lock, Thread, current_thread
from time import monotonic

from ..scheduling import (
    TaskDispatchConfig,
    TaskDispatchPolicy,
    TaskRunningActivationConfig,
    TaskRunningActivationPolicy,
    TaskSchedulingBatchSource,
    TaskWorkerAllocationConfig,
    TaskWorkerAllocationPolicy,
    WorkerServiceabilityDispatchConfig,
    WorkerServiceabilityDispatchPolicy,
)
from ..scheduling.task_scheduling_batch_source import DueTaskObservation


_LOGGER = logging.getLogger(__name__)
_TASK_BATCH_LIMIT = 100


@dataclass(frozen=True, slots=True)
class AssignmentDispatchConfig:
    worker_allocation: TaskWorkerAllocationConfig
    running_activation: TaskRunningActivationConfig
    task_dispatch: TaskDispatchConfig
    worker_allocation_interval_millis: int
    running_activation_interval_millis: int
    task_dispatch_interval_millis: int

    def __post_init__(self) -> None:
        if any(
            value <= 0
            for value in (
                self.worker_allocation_interval_millis,
                self.running_activation_interval_millis,
                self.task_dispatch_interval_millis,
            )
        ):
            raise ValueError("Dispatch lane intervals must be positive")


@dataclass(frozen=True, slots=True)
class WorkerServiceabilityDispatchLaneConfig:
    dispatch: WorkerServiceabilityDispatchConfig
    interval_millis: int

    def __post_init__(self) -> None:
        if self.interval_millis <= 0:
            raise ValueError("serviceability dispatch interval must be positive")


class _LaneId(Enum):
    RUNNING_ACTIVATION = "running-activation"
    WORKER_ALLOCATION = "worker-allocation"
    TASK_DISPATCH = "task-dispatch"
    WORKER_SERVICEABILITY = "worker-serviceability"


@dataclass(frozen=True, slots=True)
class _Lane:
    lane_id: _LaneId
    interval_millis: int
    policy: object


@dataclass(slots=True)
class _LaneRuntime:
    lane: _Lane
    inflight: bool = False
    next_eligible: float = 0.0


@dataclass(frozen=True, slots=True)
class _Completion:
    lane_id: _LaneId
    failure: BaseException | None


class DispatchConvergenceApplication:
    """Coordinate fixed Task-source dispatch lanes with single-flight batches."""

    def __init__(
        self,
        source: TaskSchedulingBatchSource,
        activation: TaskRunningActivationPolicy,
        allocation: TaskWorkerAllocationPolicy,
        dispatch: TaskDispatchPolicy,
        serviceability: WorkerServiceabilityDispatchPolicy | None,
    ) -> None:
        self.source = source
        self.activation = activation
        self.allocation = allocation
        self.dispatch = dispatch
        self.serviceability = serviceability
        self._lifecycle_lock = Lock()
        self._stop_event: Event | None = None
        self._coordinator: Thread | None = None
        self._batch_threads: set[Thread] = set()
        self._state = "STOPPED"

    @property
    def state(self) -> str:
        with self._lifecycle_lock:
            if (
                self._state == "RUNNING"
                and self._coordinator is not None
                and not self._coordinator.is_alive()
            ):
                self._state = "FAILED"
            return self._state

    def is_running(self) -> bool:
        return self.state == "RUNNING"

    def start(
        self,
        *,
        assignment: AssignmentDispatchConfig,
        serviceability: WorkerServiceabilityDispatchLaneConfig | None,
    ) -> None:
        with self._lifecycle_lock:
            if self._coordinator is not None or self._state != "STOPPED":
                raise RuntimeError("Dispatch Convergence is already started")
            stop_event = Event()
            completions: Queue[_Completion] = Queue()
            lanes = self._lanes(assignment, serviceability)
            coordinator = Thread(
                name="dispatch-convergence-coordinator",
                target=self._run,
                kwargs={
                    "stop_event": stop_event,
                    "completions": completions,
                    "lanes": lanes,
                },
                daemon=False,
            )
            self._stop_event = stop_event
            self._coordinator = coordinator
            self._state = "RUNNING"
            coordinator.start()

    def stop(self, *, timeout_millis: int) -> None:
        if timeout_millis <= 0:
            raise ValueError("stop timeout must be positive")
        with self._lifecycle_lock:
            coordinator = self._coordinator
            stop_event = self._stop_event
            if coordinator is None or stop_event is None:
                return
            self._state = "STOPPING"
            stop_event.set()
        deadline = monotonic() + timeout_millis / 1_000
        coordinator.join(timeout=max(0.0, deadline - monotonic()))
        if coordinator.is_alive():
            self._state = "FAILED"
            raise TimeoutError("Dispatch Convergence coordinator did not stop")
        for thread in tuple(self._batch_threads):
            thread.join(timeout=max(0.0, deadline - monotonic()))
        if any(thread.is_alive() for thread in self._batch_threads):
            self._state = "FAILED"
            raise TimeoutError("Dispatch Convergence batches did not stop")
        with self._lifecycle_lock:
            self._coordinator = None
            self._stop_event = None
            self._batch_threads.clear()
            self._state = "STOPPED"

    def _lanes(
        self,
        assignment: AssignmentDispatchConfig,
        serviceability: WorkerServiceabilityDispatchLaneConfig | None,
    ) -> tuple[_Lane, ...]:
        lanes = [
            _Lane(
                _LaneId.RUNNING_ACTIVATION,
                assignment.running_activation_interval_millis,
                lambda batch: self.activation.activate_running_visible_tasks(
                    batch,
                    config=assignment.running_activation,
                ),
            ),
            _Lane(
                _LaneId.WORKER_ALLOCATION,
                assignment.worker_allocation_interval_millis,
                lambda batch: self.allocation.allocate_candidate_workers(
                    batch,
                    config=assignment.worker_allocation,
                ),
            ),
            _Lane(
                _LaneId.TASK_DISPATCH,
                assignment.task_dispatch_interval_millis,
                lambda batch: self.dispatch.dispatch_tasks(
                    batch,
                    config=assignment.task_dispatch,
                ),
            ),
        ]
        if serviceability is not None:
            if self.serviceability is None:
                raise RuntimeError(
                    "serviceability lane requires its dispatch policy"
                )
            lanes.append(_Lane(
                _LaneId.WORKER_SERVICEABILITY,
                serviceability.interval_millis,
                lambda batch: self.serviceability.dispatch_probes(  # type: ignore[union-attr]
                    batch,
                    config=serviceability.dispatch,
                ),
            ))
        return tuple(lanes)

    def _run(
        self,
        *,
        stop_event: Event,
        completions: Queue[_Completion],
        lanes: tuple[_Lane, ...],
    ) -> None:
        runtimes = {
            lane.lane_id: _LaneRuntime(lane=lane)
            for lane in lanes
        }
        failed = False
        try:
            while not stop_event.is_set():
                self._drain_completions(runtimes, completions)
                self._dispatch_running(
                    runtimes,
                    completions,
                    stop_event,
                )
                self._dispatch_admission(
                    runtimes,
                    completions,
                    stop_event,
                )
                if stop_event.is_set():
                    break
                self._wait_for_work(runtimes, completions, stop_event)
        except BaseException:
            failed = True
            stop_event.set()
            _LOGGER.exception("Dispatch Convergence coordinator failed")
        finally:
            with self._lifecycle_lock:
                if self._coordinator is current_thread():
                    if failed and self._state != "STOPPING":
                        self._state = "FAILED"

    def _dispatch_running(
        self,
        runtimes: dict[_LaneId, _LaneRuntime],
        completions: Queue[_Completion],
        stop_event: Event,
    ) -> None:
        now = monotonic()
        eligible = tuple(
            runtime
            for lane_id, runtime in runtimes.items()
            if lane_id is not _LaneId.RUNNING_ACTIVATION
            and not runtime.inflight
            and now >= runtime.next_eligible
        )
        if not eligible or stop_event.is_set():
            return
        try:
            batch = self.source.acquire_running_tasks(limit=_TASK_BATCH_LIMIT)
        except Exception:
            _LOGGER.exception("Dispatch Convergence RUNNING source failed")
            for runtime in eligible:
                self._defer(runtime)
            return
        if not batch:
            for runtime in eligible:
                self._defer(runtime)
            return
        for runtime in eligible:
            self._submit(runtime, batch, completions)

    def _dispatch_admission(
        self,
        runtimes: dict[_LaneId, _LaneRuntime],
        completions: Queue[_Completion],
        stop_event: Event,
    ) -> None:
        runtime = runtimes[_LaneId.RUNNING_ACTIVATION]
        if (
            runtime.inflight
            or monotonic() < runtime.next_eligible
            or stop_event.is_set()
        ):
            return
        try:
            batch = self.source.acquire_admission_tasks(limit=_TASK_BATCH_LIMIT)
        except Exception:
            _LOGGER.exception("Dispatch Convergence ADMISSION source failed")
            self._defer(runtime)
            return
        if not batch:
            self._defer(runtime)
            return
        self._submit(runtime, batch, completions)

    def _submit(
        self,
        runtime: _LaneRuntime,
        batch: tuple[DueTaskObservation, ...],
        completions: Queue[_Completion],
    ) -> None:
        runtime.inflight = True
        thread = Thread(
            name=f"dispatch-convergence-{runtime.lane.lane_id.value}",
            target=self._execute_batch,
            args=(runtime.lane, batch, completions),
            daemon=False,
        )
        self._batch_threads.add(thread)
        try:
            thread.start()
        except BaseException:
            runtime.inflight = False
            self._batch_threads.discard(thread)
            raise

    @staticmethod
    def _execute_batch(
        lane: _Lane,
        batch: tuple[DueTaskObservation, ...],
        completions: Queue[_Completion],
    ) -> None:
        failure: BaseException | None = None
        try:
            lane.policy(batch)  # type: ignore[operator]
        except BaseException as error:
            failure = error
        finally:
            completions.put(_Completion(lane.lane_id, failure))

    def _drain_completions(
        self,
        runtimes: dict[_LaneId, _LaneRuntime],
        completions: Queue[_Completion],
    ) -> None:
        while True:
            try:
                completion = completions.get_nowait()
            except Empty:
                return
            self._apply_completion(runtimes, completion)

    def _apply_completion(
        self,
        runtimes: dict[_LaneId, _LaneRuntime],
        completion: _Completion,
    ) -> None:
        runtime = runtimes[completion.lane_id]
        runtime.inflight = False
        self._defer(runtime)
        self._batch_threads = {
            thread for thread in self._batch_threads if thread.is_alive()
        }
        if completion.failure is None:
            return
        if isinstance(completion.failure, Exception):
            _LOGGER.error(
                "Dispatch Convergence lane failed lane=%s type=%s",
                completion.lane_id.value,
                type(completion.failure).__name__,
            )
            return
        raise completion.failure

    def _wait_for_work(
        self,
        runtimes: dict[_LaneId, _LaneRuntime],
        completions: Queue[_Completion],
        stop_event: Event,
    ) -> None:
        now = monotonic()
        waits = tuple(
            max(0.0, runtime.next_eligible - now)
            for runtime in runtimes.values()
            if not runtime.inflight
        )
        timeout = min(waits, default=0.05)
        timeout = min(timeout, 0.05)
        try:
            completion = completions.get(timeout=timeout)
        except Empty:
            stop_event.wait(0)
            return
        self._apply_completion(runtimes, completion)

    @staticmethod
    def _defer(runtime: _LaneRuntime) -> None:
        runtime.next_eligible = (
            monotonic() + runtime.lane.interval_millis / 1_000
        )
